lib/Sema/Constraints.cpp

```diff
@@ -0,0 +1,464 @@
+    , _overload { type }
+    , _locator(locator)
+{
+    std::copy(
+        typeVars.begin(), typeVars.end(), getTypeVariablesBuffer().begin()
+    );
+    *getTrailingObjects<glu::ast::FunctionDecl *>() = choice;
+}
+
+/// Recursively gather the set of type variables referenced by this constraint.
+static void gatherReferencedTypeVars(
+    Constraint *constraint,
+    llvm::SmallPtrSetImpl<glu::types::TypeVariableTy *> &typeVars
+)
+{
+    switch (constraint->getKind()) {
+    case ConstraintKind::Disjunction:
+        for (auto nested : constraint->getNestedConstraints())
+            gatherReferencedTypeVars(nested, typeVars);
+        return;
+
+    case ConstraintKind::Conjunction:
+        typeVars.insert(
+            constraint->getTypeVariables().begin(),
+            constraint->getTypeVariables().end()
+        );
+        return;
+    case ConstraintKind::Bind:
+    case ConstraintKind::BindToPointerType:
+    case ConstraintKind::ArgumentConversion:
+    case ConstraintKind::Conversion:
+    case ConstraintKind::OperatorArgumentConversion:
+    case ConstraintKind::CheckedCast:
+    case ConstraintKind::Equal:
+    case ConstraintKind::UnresolvedValueMember:
+    case ConstraintKind::ValueMember:
+    case ConstraintKind::Defaultable:
+    case ConstraintKind::FallbackType:
+    case ConstraintKind::ExplicitGenericArguments:
+    case ConstraintKind::LValueObject:
+        typeVars.insert(
+            llvm::cast<glu::types::TypeVariableTy>(constraint->getFirstType())
+        );
+        typeVars.insert(
+            llvm::cast<glu::types::TypeVariableTy>(constraint->getSecondType())
+        );
+        break;
+
+    case ConstraintKind::BindOverload:
+        typeVars.insert(
+            llvm::cast<glu::types::TypeVariableTy>(constraint->getFirstType())
+        );
+
+        // // Special case: the base type of an overloading binding.
+        if (constraint->getKind() == ConstraintKind::BindOverload) {
+            typeVars.insert(
+                llvm::cast<glu::types::TypeVariableTy>(
+                    constraint->getOverloadChoice()->getType()
+                )
+            );
+        }
+
+        break;
+
+    case ConstraintKind::SyntacticElement:
+        typeVars.insert(
+            constraint->getTypeVariables().begin(),
+            constraint->getTypeVariables().end()
+        );
+        break;
+    }
+}
+
+Constraint *Constraint::create(
+    llvm::BumpPtrAllocator &allocator, ConstraintKind kind,
+    glu::types::Ty first, glu::types::Ty second, glu::ast::ASTNode *locator,
+    llvm::ArrayRef<glu::types::TypeVariableTy *> extraTypeVars
+)
+{
+    llvm::SmallPtrSet<glu::types::TypeVariableTy *, 4> typeVars;
+
+    assert(first && "First type is Null");
+    assert(second && "Second type is Null");
+    assert(locator && "Locator is Null");
+
+    if (first->getKind() == glu::types::TypeKind::TypeVariableTyKind)
+        typeVars.insert(llvm::cast<glu::types::TypeVariableTy>(first));
```

Comment:
never check the kind directly... use llvm::isa or
```suggestion
    if (auto *ty = llvm::dyn_cast<glu::types::TypeVariableTy>(first))
        typeVars.insert(ty);
```